nvflare/fuel/f3/communicator.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2021-2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import os
+from typing import Optional
+
+from nvflare.fuel.f3 import drivers
+from nvflare.fuel.f3.comm_error import CommError
+from nvflare.fuel.f3.drivers.driver import Driver
+from nvflare.fuel.f3.drivers.driver_manager import DriverManager
+from nvflare.fuel.f3.drivers.driver_params import DriverParams
+from nvflare.fuel.f3.drivers.net_utils import parse_url
+from nvflare.fuel.f3.endpoint import Endpoint, EndpointMonitor
+from nvflare.fuel.f3.message import Message, MessageReceiver
+from nvflare.fuel.f3.sfm.conn_manager import ConnManager, Mode
+
+log = logging.getLogger(__name__)
+
+
+class Communicator:
+    """FCI main communication API"""
+
+    driver_mgr = DriverManager()
+    # Load all the drivers in the drivers module
+    driver_mgr.register_folder(os.path.dirname(drivers.__file__), drivers.__package__)
+
+    def __init__(self, local_endpoint: Endpoint):
+        self.local_endpoint = local_endpoint
+        self.monitors = []
+        self.conn_manager = ConnManager(local_endpoint)
+
+    def start(self):
+        """Start the communicator and establishing all the connections
+
+        Raises:
+            CommError: If any error encountered while starting up
+        """
+        self.conn_manager.start()
+        log.debug(f"Communicator for local endpoint: {self.local_endpoint.name} is started")
+
+    def stop(self):
+        """Stop the communicator and shutdown all the connections
+
+        Raises:
+            CommError: If any error encountered while shutting down
+        """
+        self.conn_manager.stop()
+        log.debug(f"Communicator for local endpoint: {self.local_endpoint.name} has stopped")
+
+    def register_monitor(self, monitor: EndpointMonitor):
+        """Register a monitor for endpoint lifecycle changes
+
+        This monitor is notified for any state changes of all the endpoints.
+        Multiple monitors can be registered.
+
+        Args:
+            monitor: The class that receives the endpoint state change notification
+
+        Raises:
+            CommError: If any error happens while sending the request
+        """
+        self.conn_manager.add_endpoint_monitor(monitor)
+
+    def get_endpoint(self, name: str) -> Optional[Endpoint]:
+        """Find endpoint by name
```

Comment:
should we just rename this method to `find_endpoint` then?

Reply:
Good idea. Done.